Mathlib/LinearAlgebra/Span.lean

```diff
@@ -150,6 +150,24 @@ theorem span_induction {p : M → Prop} (h : x ∈ span R s) (Hs : ∀ x ∈ s,
   ((@span_le (p := ⟨ ⟨⟨p, by intros x y; exact H1 x y⟩, H0⟩, H2⟩)) s).2 Hs h
 #align submodule.span_induction Submodule.span_induction
 
+open AddSubmonoid in
+theorem span_eq_closure {s : Set M} : (span R s).toAddSubmonoid = closure (@univ R • s) := by
+  refine le_antisymm (fun x hx ↦ span_induction hx
+    (fun x hx ↦ subset_closure ⟨1, x, trivial, hx, one_smul R x⟩)
+    (zero_mem _) (fun _ _ ↦ add_mem) fun r m hm ↦ closure_induction hm ?_ ?_ ?_) (closure_le.2 ?_)
+  · rintro _ ⟨r, m, -, hm, rfl⟩; exact smul_mem _ _ (subset_span hm)
+  · rintro _ ⟨r', m, -, hm, rfl⟩; exact subset_closure ⟨r * r', m, trivial, hm, mul_smul r r' m⟩
+  · rw [smul_zero]; apply zero_mem
+  · intro m m' hm hm'; rw [smul_add]; exact add_mem hm hm'
+
+@[elab_as_elim]
+theorem closure_induction {p : M → Prop} (h : x ∈ span R s) (H0 : p 0)
+    (H1 : ∀ x y, p x → p y → p (x + y)) (H2 : ∀ a : R, ∀ x ∈ s, p (a • x)) : p x := by
+  rw [← mem_toAddSubmonoid, span_eq_closure] at h
+  refine AddSubmonoid.closure_induction h ?_ H0 H1
+  rintro _ ⟨r, m, -, hm, rfl⟩
+  exact H2 r m hm
```

Comment:
Isn't this `span_induction`?

Reply:
This one only need the R-multiples of elements in the generating set to satisfy the predicate, rather than R-multiples of all elements satisfying the predicate. The conditions are easier to verify when proving `span_smul_of_span_eq_top`. (I'm not aware of a way proving it without the commutativity assumption without going through `closure_induction`.)

Reply:
Maybe put your explanation into a docstring?